Add only TLB support

yechen3 opened the pull request:
Extracted TLB-only features from the UVM support in PR#108 and integrated them into existing classes to minimize API changes. Current features include:
	1.	A 4096-entry, fully associative TLB with LRU replacement policy by default
	2.	4-port TLB capable of completing up to 4 lookups per cycle
	3.	Support for 4KB pages, with a page table walk latency of 100 cycles
	4.	All parameters are configurable

To-do-list:
	1.	Need a mechanism to disable TLB (i.e., bypass ldst_unit::tlb_cycle)
	2.     Probably need to merge requests from TLB to GMMU of same virtual pages to reduce traffic.

tgrogers commented:
@William-An , @FJShen - can you guys look at this? This is a lot of code.

tgrogers commented on .github/workflows/accelsim.yml:
```diff
@@ -16,7 +16,7 @@ on:
 env:
   ACCELSIM_REPO: https://github.com/purdue-aalp/accel-sim-framework-public.git
-  ACCELSIM_BRANCH: dev
+  ACCELSIM_BRANCH: dev-tlb
```
Do we need the dev-tlb branch of accel-sim for gpgpu-sim to work?
If the config is not configured to use tlbs, then does it still need this branch?

William-An replied:
I think this is a circular dependency issue: the tlb version of Accel-Sim needs the tlb version of gpgpu-sim and vice versa. We should have some ways to deal with this. Some ideas I have:
- Specify the branch to use for CI runs? Not sure if github allows inputs from users.
- First use the `dev` branch to test `Accel-Sim/dev` with `GPGPU-Sim/dev-tlb`, make sure things are not broken. Then add additional CI test for `Accel-Sim/dev-tlb` and `GPGPU-Sim/dev-tlb`

Also, is it possible to control whether to enable TLB functionality or not?

yechen3 replied:
The only chnage in dev-tlb branch is adding m_memory_stats argument in trace_shader_core_ctx() within trace_driven.cc. It still needs this change regardless of whether tlb config is used or not. So is it better to overload this functions and create two variants.

tgrogers commented on src/abstract_hardware_model.cc:
```diff
@@ -178,6 +178,18 @@ void gpgpu_functional_sim_config::ptx_set_tex_cache_linesize(
   m_texcache_linesize = linesize;
 }
 
+// void gpgpu_functional_sim_config::convert_byte_string() {
```
why are we commenting out code?
If you don't need it remove it.

FJShen replied:
Please remove or uncomment

tgrogers commented on src/gpgpu-sim/mem_latency_stat.cc:
```diff
@@ -183,6 +183,35 @@ memory_stats_t::memory_stats_t(unsigned n_shader,
+  mf_page_hit = (unsigned long long *)calloc(num_cluster, sizeof(unsigned long long));
+  mf_page_miss = (unsigned long long *)calloc(num_cluster, sizeof(unsigned long long));
+
+  // mf_page_fault_outstanding = 0;
```
commented code - why?

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -448,9 +448,9 @@ void shader_core_ctx::create_exec_pipeline() {
   }
 
-  m_ldst_unit = new ldst_unit(m_icnt, m_mem_fetch_allocator, this,
+  m_ldst_unit = new ldst_unit(m_gpu, m_icnt, m_mem_fetch_allocator, this,
```
the ldst unit now needs the whole GPU because it needs to look up some global page table information?

yechen3 replied:
No, it needs the whole GPU to register tlb flush callback functions. The gpu page table is managed inside gmmu class, which is instantiated once per gpu. Additionally, I'm assuming the page table will never be missed (as in regular memcpy and UVA setups), and each page table walk incurs a constant latency of 100 cycles.

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2011,7 +2013,7 @@ mem_stage_stall_type ldst_unit::process_cache_access(
   if (status == HIT) {
     assert(!read_sent);
-    inst.accessq_pop_back();
+    inst.accessq_pop_front();
```
Why are we changing this to FIFO?

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2257,18 +2259,135 @@ bool ldst_unit::texture_cycle(warp_inst_t &inst, mem_stage_stall_type &rc_fail,
+bool ldst_unit::tlb_cycle(warp_inst_t &inst,
+                          mem_stage_stall_type &stall_reason,
+                          mem_stage_access_type &access_type) {
+  printf("tlb_cycle\n");
```
Change to use the TRACE system. Add an entry to VMEM_SYS

William-An replied:
Speaking of TRACE, I am thinking of using a more standardized approach, i.e. some official logging C++ libraries like [spdlog](https://github.com/gabime/spdlog), with modern C++ formatting support like [fmt](https://github.com/fmtlib/fmt).

This should give a "python-like" logging experience to facilitate coding.

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2257,18 +2259,135 @@ bool ldst_unit::texture_cycle(warp_inst_t &inst, mem_stage_stall_type &rc_fail,
                              mem_stage_stall_type &stall_reason,
                              mem_stage_access_type &access_type) {
+  mem_stage_stall_type stall_cond = NO_RC_FAIL;
+  // if (!inst.empty())
```
commented code

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2257,18 +2259,135 @@ bool ldst_unit::texture_cycle(warp_inst_t &inst, mem_stage_stall_type &rc_fail,
+      stall_reason = TLB_STALL;
+      return false;
+    } else {
+      // printf("GMMU CU Queue\n");
```
this seems useful. This is why we have traces. Add it to the trace.

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2257,18 +2259,135 @@ bool ldst_unit::texture_cycle(warp_inst_t &inst, mem_stage_stall_type &rc_fail,
+        }
+      }
+
+      // printf("pushing back from cu queue\n");
```
trace

tgrogers commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2618,20 +2744,24 @@ void ldst_unit::init(mem_fetch_interface *icnt,
   m_last_inst_gpu_sim_cycle = 0;
   m_last_inst_gpu_tot_sim_cycle = 0;
+
+  gpu->getGmmu()->register_tlbflush_callback(
```
ok - the global MMU gets a call from every SM?
This should really be done at a different level. I don't like pushing this global stuff down into the ldst unit.

yechen3 replied:
This is only called during initialization to register TLB flush callback functions. These callbacks should be triggered when the TLB is flushed—for example, when a kernel completes.

tgrogers reviewed (changes requested):
Overall - I think we need to have a meeting with @William-An @LAhmos @JRPan @christindbose  to discuss this.

Generally, the code has some small issues I would like to see fixed - but I would like you @yechen3 to create a presentation / readme that describes the feature to us so we are all on the same page with what is happening.

Also, some basic questions:
 1 - do the config files have to change?
 2 - is there any slowdown in sim time?
 3 - does the change effect correlation?

Also please take the addition to the trace system seriously - this is invaluable to debug.

beneslami commented:
I have been actively and eagerly following this PR. 
So far, no concrete verdict for this branch ?

If the functionality works fine, I intend to manually add these changes to my own custom AccelSim.

yechen3 commented:
> Overall - I think we need to have a meeting with @William-An @LAhmos @JRPan @christindbose to discuss this.
> 
> Generally, the code has some small issues I would like to see fixed - but I would like you @yechen3 to create a presentation / readme that describes the feature to us so we are all on the same page with what is happening.
> 
> Also, some basic questions: 1 - do the config files have to change? 2 - is there any slowdown in sim time? 3 - does the change effect correlation?
> 
> Also please take the addition to the trace system seriously - this is invaluable to debug.

Ok, I will prepare some slides to present next week and work on fixing those issue this week.

Answers to @tgrogers questions: 
1. I added three configurable variables—_tlb_size_, _page_size_, and _page_table_walk_latency_. They all have default values, so existing config files don’t need to be modified.
2. There’s no noticeable slowdown for the GPU-Micro benchmarks or Rodinia-3.1 workloads.
3. The correlations (gpc_cycles) remain largely consistent:
 **Rodinia-3.1:** QV100-SASS
    (17 apps: 1 < 1% Err, 6 under, 11 over, 4 < 10% Err)
    [Correl = 0.9836 | Err = 27.83% | Agg_Err = 27.84% | RPD = 25.20% | NMSE = 0.45]
**GPU-Microbenchmark:** QV100-SASS
    (11 apps: 0 < 1% Err, 6 under, 5 over, 2 < 10% Err)
    [Correl = 0.9419 | Err = 98.93% | Agg_Err = 30.09% | RPD = 38.08% | NMSE = 0.50]

yechen3 commented:
> I have been actively and eagerly following this PR. So far, no concrete verdict for this branch ?
> 
> If the functionality works fine, I intend to manually add these changes to my own custom AccelSim.

@beneslami Thanks for following the PR! The current version should work, but it hasn’t been fully verified yet. We’re planning to have a meeting to discuss and review everything in more detail. I’ll make sure to update you with the outcome afterward.

FJShen commented on src/gpgpu-sim/gpu-sim.cc:
```diff
@@ -322,6 +323,12 @@ void memory_config::reg_options(class OptionParser *opp) {
+  option_parser_register(
+      opp, "-page_table_walk_latency", OPT_INT64, &page_table_walk_latency,
+      "Average page table walk latency (in core cycle).", "100");
+  option_parser_register(opp, "-page_size", OPT_CSTR, &page_size_string,
```
This argument seems unused in the whole PR. Is it used anywhere? If used, please consider implementing safety check code to test its validity; if not, please consider removing it.

yechen3 replied:
It's used inside gmmu_t constructor to decide the page number from an address. What do you mean by the validity?

FJShen replied:
The description says "GDDR page size, only 4KB/2MB avaliable." It would be nice to have a sanity check in the code for this.

FJShen commented on src/abstract_hardware_model.h:
```diff
@@ -559,6 +559,8 @@ class gpgpu_functional_sim_config {
   int get_checkpoint_CTA_t() const { return checkpoint_CTA_t; }
   int get_checkpoint_insn_Y() const { return checkpoint_insn_Y; }
 
+  // void convert_byte_string();
```
Please remove or uncomment

FJShen commented on src/abstract_hardware_model.h:
```diff
@@ -578,6 +580,10 @@ class gpgpu_functional_sim_config {
+
+  // protected:
+  // int page_size;
+  // char *page_size_string;
```
Please remove or uncomment.

FJShen reviewed (changes requested):
Please consider adding a block of comments to provide an overview of TLB feature. How it works, what it interacts with, its assumptions, etc.

JRPan commented:
> Ok, I will prepare some slides to present next week and work on fixing those issue this week.

Please try to accommodate Western Time Zone :) 
Thanks.

JRPan commented on src/gpgpu-sim/shader.cc:
```diff
@@ -2257,18 +2259,135 @@ bool ldst_unit::texture_cycle(warp_inst_t &inst, mem_stage_stall_type &rc_fail,
+bool ldst_unit::tlb_cycle(warp_inst_t &inst,
+                          mem_stage_stall_type &stall_reason,
+                          mem_stage_access_type &access_type) {
+  printf("tlb_cycle\n");
```
This should just be removed.

JRPan commented:
Copilot has some valid points

yechen3 commented:
> Generally, the code has some small issues I would like to see fixed - but I would like you @yechen3 to create a presentation / readme that describes the feature to us so we are all on the same page with what is happening.

@tgrogers @JRPan @William-An @FJShen I've made some slides in [here](https://purdue0.sharepoint.com/sites/RogersLab/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2FRogersLab%2FShared%20Documents%2FPresentations) (gpgpu-sim-tlb.pptx). Please let me know when you guys are available to meet or if you have any questions.

beneslami commented:
Hi @yechen3 
I'm curious and excited to know about the latest status of your conclusion regarding TLB implementation in AccelSim. 

Also, I would like to highlight a few points. Maybe these points are already taken care of or they may be insightful for you:

1- I went through your code and saw that the class GMMU is implemented. I was wondering if it is possible to simulate the presence of IOMMU by tweaking the internal attributes of GMMU class ? As you know, CPU-GPU executions use unified virtual memory and each time a TLB miss happens in Last Level TLB of GPU, it sends request to IOMMU. This means we are supposed to experience some amount of PCI-e latency (Based on [this paper](https://ieeexplore.ieee.org/abstract/document/10609639)). 

2- It's good that GMMU is defined as a class because in my customized AccelSim, I implemented chiplet-based GPU system. And I can create GMMU object per chiplet. The TLB is also implemented class-based ? As far as I remember, the TLB was defined as a global map (correct me if I'm wrong)? It's a scalable idea to have a base TLB, then L1 TLB is defined, inheriting base TLB, and so on. It's because address translation is not yet scalable in MCM-GPU systems and there needs to have different micro-architectural tweaks and analysis, which requires agile baseline implementation of TLB.

Thanks 
Ben

yechen3 commented:
> Hi @yechen3 I'm curious and excited to know about the latest status of your conclusion regarding TLB implementation in AccelSim.
> 
> Also, I would like to highlight a few points. Maybe these points are already taken care of or they may be insightful for you:
> 
> 1- I went through your code and saw that the class GMMU is implemented. I was wondering if it is possible to simulate the presence of IOMMU by tweaking the internal attributes of GMMU class ? As you know, CPU-GPU executions use unified virtual memory and each time a TLB miss happens in Last Level TLB of GPU, it sends request to IOMMU. This means we are supposed to experience some amount of PCI-e latency (Based on [this paper](https://ieeexplore.ieee.org/abstract/document/10609639)).
> 
> 2- It's good that GMMU is defined as a class because in my customized AccelSim, I implemented chiplet-based GPU system. And I can create GMMU object per chiplet. The TLB is also implemented class-based ? As far as I remember, the TLB was defined as a global map (correct me if I'm wrong)? It's a scalable idea to have a base TLB, then L1 TLB is defined, inheriting base TLB, and so on. It's because address translation is not yet scalable in MCM-GPU systems and there needs to have different micro-architectural tweaks and analysis, which requires agile baseline implementation of TLB.
> 
> Thanks Ben

Hi @beneslami, we are going to have an internal review next Monday. Regarding your suggestions:

1. We adopted TLB implementation from [UVMSmart](https://github.com/DebashisGanguly/gpgpu-sim_UVMSmart), which simulates the full virtual memory system, (i.e. PCIe queue, page table, prefetcher). However, the simulation time was significantly increased as a result. That’s the main reason we decided to simplify the design and retain only the TLB component. Also, in most scenarios, CPU-GPU memory coherence (as in UVM) is uncommon, while GPU peer memory access via Unified Virtual Addressing (UVA) is more typical.

2. It's a good suggestion, we’ll go ahead and make that change.

Thanks again for your feedback.

William-An commented on src/gpgpu-sim/gpu-sim.cc:
```diff
@@ -93,6 +93,7 @@ tr1_hash_map<new_addr_type, unsigned> address_random_interleaving;
 #define L2 0x02
 #define DRAM 0x04
 #define ICNT 0x08
+#define GMMU 0x10
```
Minor nitpicking: Can you rewrite this in the form of `1 << 1, 1 << 2`? Just to keep things a bit cleaner.

FJShen replied:
If we were to adopt the form "1 << n", we better wrap them in a pair of parentheses at macro definition. The C++ shift operators have low precedence.

William-An commented on src/gpgpu-sim/gpu-sim.h:
```diff
@@ -521,6 +528,60 @@ class gpgpu_sim_config : public power_config,
+  class memory_stats_t *m_memory_stats;
+};
+  
+struct lp_tree_node {
```
Is this struct ever used?

William-An commented on src/gpgpu-sim/gpu-sim.h:
```diff
@@ -521,6 +528,60 @@ class gpgpu_sim_config : public power_config,
+  // page table walk delay queue
+  std::list<page_table_walk_latency_t> page_table_walk_queue;
+
+  enum class latency_type {
```
Same for this enum, is it ever used?

William-An commented on src/gpgpu-sim/shader.h:
```diff
@@ -1466,11 +1486,23 @@ class ldst_unit : public pipelined_simd_unit {
+  std::list<mem_fetch *> m_cu_gmmu_queue;
+
+  // set of virtual addresses present in TLB
+  std::list<mem_addr_t> tlb;
```
I guess we can document somewhere that the TLB is a fully-associative cache.

William-An commented on src/gpgpu-sim/shader.h:
```diff
@@ -1366,6 +1367,9 @@ class ldst_unit : public pipelined_simd_unit {
   void fill(mem_fetch *mf);
+  // function to fill the gmmu to cu queue
+  // from the cluster to load/store unit
+  void fill_mem_access(mem_fetch *mf);
```
If this method is only used for TLB-related actions, we should give a specific name rather than a generic "mem_access".

William-An commented on src/gpgpu-sim/shader.h:
```diff
@@ -1466,11 +1486,23 @@ class ldst_unit : public pipelined_simd_unit {
   unsigned long long m_last_inst_gpu_tot_sim_cycle;
 
+  // two queues that interface with texture processor cluster
+  std::list<mem_fetch *> m_gmmu_cu_queue;
```
These two queues are the core-specific queues. It will be nice if they have different names from the cluster queues.

William-An commented on src/gpgpu-sim/shader.h:
```diff
@@ -2670,6 +2731,11 @@ class simt_core_cluster {
+
+  // queues that pass memory accesses between core and GMMU
+  // as cluster interfaces between CU and GMMU
+  std::list<mem_fetch *> m_gmmu_cu_queue;
```
We should use different names to differentiate the cluster queue and core queue.

Something like `m_mmu_gmmu2cluster_queue` and `m_mmu_cluster2core_queue`

beneslami commented:
@yechen3 @William-An 
Hi guys, 

Is there any update from the TLB project ?
